### PR TITLE
Move jsonlog read logic

### PR DESCRIPTION
**Pull request by cpuguy83**
First refactors the log writer object to be specifically about writing `logger.Message`'s rather than a generic io.Writer. This allows us to, in the future, make optimizations for reads.

Then moves the file reading logic from the json logger into the new actual log file object.
This allows the read logic to be re-used for https://github.com/moby/moby/issues/33475 and potentially also daemon logging.

This should also simplify locking strategies for compressing rotated files.

jsonfile is now only responsible for encoding/decoding `logger.Message`'s and passing off to the log file object.

### CODE REVIEW

**Review by miaoyq — COMMENTED** [review]
`rotaterfilewriter.go` contains both write and read logic , so I think we should name `rotaterfilewriter.go` to `logfile.go` or other appropriate name.

**Comment by coolljt0725** [comment]
need a rebase

**Comment by cpuguy83** [comment]
Rebased

File: daemon/logger/loggerutils/logfile.go
Comment (dperny) [inline daemon/logger/loggerutils/logfile.go]: i can see an error occurring here where the docker daemon dies during a log rotation, so there's a number skipped, and then the next time you try to rotate logs, an error occurs and aborts the rotation, which leaves you in a bad state until you totally wipe out all the logs. do you see this as well?

**Review by dperny — COMMENTED** [review]
I see some maybeissues in the `daemon/logger/loggerutils/logfile.go` that might be a problem, but I realized it's old code so they're probably either fine or out of scope to fix at this time. I can go back and add them if you want. LGTM otherwise, but I Am Not A Maintainer.

File: daemon/logger/loggerutils/logfile.go
Comment (cpuguy83) [inline daemon/logger/loggerutils/logfile.go]: I'm actually not seeing it since this is skipping errors when the error is a `NotExist` error.

**Comment by thaJeztah** [comment]
looks like this needs another rebase

also ping @dperny PTAL

**Comment by cpuguy83** [comment]
Rebased

**Comment by dperny** [comment]
LGTM, but i am not a maintainer.

**Review by AkihiroSuda — APPROVED** [review]
LGTM

**Comment by AkihiroSuda** [comment]
@coolljt0725 PTAL?

**Comment by coolljt0725** [comment]
LGTM  ping @vieux

**Comment by thaJeztah** [comment]
ping @cpuguy83 looks like we have two LGTM's, but needs a rebase

**Comment by thaJeztah** [comment]
@vieux gave an LGTM offline (after rebase)

**Comment by cpuguy83** [comment]
Seems to be a legit windows error. I'll have to dig through what change caused it.

**Comment by cpuguy83** [comment]
All green after a rebase.

**Comment by thaJeztah** [comment]
Sorry @cpuguy83 - looks like this needs another rebase

**Comment by cpuguy83** [comment]
Oh I knew that was coming :)

**Comment by cpuguy83** [comment]
Rebased and green.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by yongtang** [comment]
Seems we have enough LGTMs, merging.

**Pull request merged**
